UNIX/SOURCE/xkal-1.13/xkal.man
.\"
.\"	Xkal : Graphical appointment calendar
.\"
.\"	George Ferguson, ferguson@cs.rochester.edu,  27 Oct 1990.
.\"	Version 1.1 - 27 Feb 1991.
.\"
.\"	$Id: xkal.man,v 2.2 91/03/13 13:36:09 ferguson Exp $
.\"
.TH XKAL 1 "26/10/90"
.ds ]W U of Rochester
.SH NAME
xkal \- graphical appointment calendar
.SH SYNOPSIS
.B xkal
[X\ Toolkit\ options]
[-appoints\ filename]
[-date\ date]
[-listOnly]
[-silent]
[-numMonths\ 1|2|3|12]
[-[no]exitUsesLevels]
[-[no]bothShown]
[-[no]opaqueDates]
[-[no]dowLabels]
[-[no]titlebar]
[-version]
[-help]
.SH DESCRIPTION
.PP
Xkal is a graphical appointment calendar that allows you to maintain
many types of reminders, displays them in several different formats,
and allows you to add and edit them. Silent and non-graphic modes are
provided to allow xkal to be used to check for appointments without
interaction.
.PP
The general behaviour of xkal is to display one or more months. Clicking
Button1 on a particular day displays that day's reminder information,
either next to the month or in a popup window (see the
.B bothShown
resource, below). The reminders can be edited within this window, and
changes are automatically saved when the tool is exited. Changes can be
aborted or the appointments can be saved immediately using other actions.
Most available actions are provided by a set of menuButtons above the month
display. Clicking and holding on these buttons displays a menu from which the
desired action can be selected. (See below under Translation Actions for a
list of the actions and ways to customize xkal's behaviour.)
.PP
Reminders can come from two different sources:
.B system
reminders cannot be edited or deleted;
.B personal
reminders can. See below under OPTIONS, FILE FORMATS, and FILES for more
details about these files. When appointments conflict (ie. there is more
than one reminder that applies for a given day and time), xkal moves one
of them to a different time (see the
.B rearrangeSilently
resource).
.PP
Each reminder can have a "criticality level" associated with it, indicating
its importance. This is reflected in the month display by shading (on
monochrome displays) or coloring (on color displays) days depending on the
total criticality of their appointments. Typically, frequent or regular
reminders would be given level zero, and only more important or unique
reminders would influence the display (although they would all be displayed
when the day is selected).
.PP
Double-clicking on a reminder will pop up a window in which the entry can
be edited, changing things such as the criticality level as desired. Other
fields can also be changed, for example to make the reminder weekly or
monthly rather than for that specific day.
Clicking on the applyButton adds the appointment, the revertButton resets
the text items but does NOT undo the effects of previous "apply"s, the
deleteButton deletes any reminders matching the given information exactly,
and the dismissButton pops down the window.
.PP
Xkal is designed to be completely customizable. The section below,
CUSTOMIZING XKAL, describes the widget hierarchies and other relevant
resources to help with this. In particular, the relative positions of most
of xkal's main components is completely flexible. This version of xkal
also supports full customization of names for use with languages other than
English.
.SH OPTIONS
.PP
The following non-widget resources can be set on the command line as well
as in a resource file. As usual, when given on the command line they can be
abbreviated to their shortest unique prefix, often the first letter.
Furthermore, xkal accepts all the standard X Toolkit options.
.IP "\fB-appoints\fP filename"
Uses the given filename as the name of the personal appointment file.
The format of this file is described below under FILE FORMATS. A leading
tilde (`~') is expanded to the value of $HOME. The default is `~/.appoints'.
This option corresponds to the resource
.BR personalAppoints .
.IP "\fB-date\fP date"
Changes xkal's notion of the current date (usually obtained from the
system). The date string should be as described below under DATE FORMATS,
with whitespace escaped if given on the command line.
.IP "\fB-listOnly\fP"
If given, then xkal simply lists the appointments and weekly reminders
to stdout, and exits without interaction. The exit status is the number of
appointments for the day. See also
.BR exitUsesLevels ,
below.
.IP "\fB-silent\fP"
Like above, but simply sets the exit status and doesn't print anything.
Useful for checking whether there are any appointments. See also
.BR -exitUsesLevels ,
below.
.IP "\fB-exitUsesLevels\fP"
If given, then the exit status returned in
.B listOnly
and
.B silent
modes is the total of the criticality levels of the day's appointments,
rather than the number of them.
.IP "\fB-noexitUsesLevels\fP"
The opposite of
.BR -exitUsesLevels.
The exit status is the number of appointments, rather than their
criticality total. This is the default.
.IP "\fB-numMonths\fP 1|3|12"
Determines whether xkal will start displaying one, three, or twelve months.
This can be changed while xkal is running using the
.B xkal-numMonths()
action described below under Translation Actions. The default is to show one
month.
.IP "\fB-bothShown\fP"
If given, then the day's appointments are displayed in the main tool
window, otherwise a popup window is created for them. When a day window is
popped up, it becomes the selected window and subsequent appointment updates
are done in it. It can be selected and deselected using the buttons. If no
popup window is selected when a day is clicked on, then a new popup window
will be created. The
.B Dismiss
button destroys the popup window. When this option is given (the default), no
popup windows are ever created.
.IP "\fB-nobothShown\fP"
The opposite of
.BR -bothShown .
.IP "\fB-opaqueDates\fP"
If given, then the numbers in the dayButtons are drawn opaquely, which
some people find easier to read when the day is stippled. This is the
default.
.IP "\fB-noopaqueDates\fP"
The opposite of
.BR -opaqueDates .
.IP "\fB-dowLabels\fP"
If given, labels are provided at the top of each column in each month
indicating the day of the week.
.IP "\fB-nodowLabels\fP"
The opposite of
.BR -dowLabels .
The default is
.BR nodowLabels .
.IP "\fB-titlebar\fP"
If true, the titlebar of the main window and any popup day windows is
changed to reflect the selected day. This is the default. This option
corresponds to the resource
.BR useTitlebar .
.IP "\fB-notitlebar\fP"
The opposite of
.BR -useTitlebar .
.IP "\fB-updateInterval\fP mins"
The number of minutes between automatic checkpointing (saving) of the
appointments. If this is zero, then no automatic checkpointing is done.
The default is to checkpoint every five minutes.
.IP "\fB-version\fP"
Prints the current version of xkal to stdout and exits.
.IP "\fB-help\fP"
Prints the usage message, summarizing these options.
.SH "CUSTOMIZING XKAL"
.PP
Xkal has a default set of resources built in. If you wish to customize the
tool, take a copy of the default application defaults file (see
FILES below) and modify it. Then, before invoking xkal, set the environment
variable
.B XAPPLRESDIR
to the directory containing your private copy. Alternatively, you can place
entries in your .Xdefaults file or provide them with the -xrm toolkit
option.
.SH "Other Non-Widget Resources"
.PP
The following resources customize xkal's file operations:
.IP "\fBsystemAppoints\fP"
The file for system (non-editable) reminders. Default is to not have a
system reminder file.
.IP "\fBbackupExtension\fP"
The suffix used when the personal appointments file is backed up before
saving it. If this is empty then the file is not backed up. The default is
to append a tilde (`~').
.IP "\fBoutputFormat\fP"
This resource specifies the desired "canonical" output format. It should
be a string with printf-style escapes that are replaced with parts of the
reminder as follows:
.nf
.na
	%w	short form of day of the week
	%W	long form of day of the week
	%d	day
	%m	short form of month
	%M	long form of month
	%n	numeric form of month
	%y	year
	%Y	year modulo 100
	%t	time
	%l	criticality level
	%~	space iff previous pattern was printed
	%/	slash iff previous pattern was printed
.ad
.fi
The default is "%w%~%d%~%m%~%y%~%t%~%l%~". Another possibility might be
"~%n%/%d%/%Y%~%w%~%t%~%l%~" (see
.BR daySlashMonth ,
below).
.IP "\fBdaySlashMonth\fP"
This resource controls how dates with slashes are interpreted. If it is
True, then these dates are interpreted as "day/month" or "day/month/year"
rather than the default "mon/day" or "mon/day/year".
.IP "\fBlevelDelim\fP"
This resource specifies the characters that, when encountered while parsing
a reminder, indicate its criticality level. If it is one character, then
that character is used for both the opening and the closing delimiter. The
default is to have the criticality level enclosed in ampersands (`@').
.IP "\fBdefaultLevel\fP"
The default criticality level for newly-created appointments. This can be
changed by editing the appointment. The default is one (`1').
.PP
The following resources customize various aspects of dates:
.IP "\fBlongDowStrings\fP"

UNIX/SOURCE/xkal-1.13/xkal.c
/*
 *	xkal.c : graphical appointment tool
 *
 *	George Ferguson, ferguson@cs.rochester.edu,  27 Oct 1990.
 *	Version 1.1 - 27 Feb 1991.
 *
 *	$Id: xkal.c,v 2.2 91/02/28 11:23:24 ferguson Exp $
 *
 */
#ifndef lint
static char ident[] = "@(#) xkal version 1.1 - ferguson@cs.rochester.edu";
#endif
#include <stdio.h>
#include <ctype.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xmu/Drawing.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/MenuButton.h>
#include <X11/Xaw/Cardinals.h>	
#include "EzMenu.h"
#include "app-resources.h"
#include "resources.h"
#include "month.h"
#include "day.h"
#include "edit.h"
#include "edit-defaults.h"
#include "db.h"
#include "util.h"
#include "date-strings.h"
#include "patchlevel.h"
#ifdef USE_ALERT
#include "alert.h"
#endif

/*	-	-	-	-	-	-	-	-	*/
/*
 * Functions defined in this file:
 */
int main();
void quit(),quit_nosave(),save(),next(),prev(),today(),setNumMonths();
void focusThisItem(), focusNoItem(), focusNextItem(), focusPrevItem();
void escapeToSystem();
void timeoutProc();

static void initGraphics(), initGCs(), initWidgets(), createMonthFormDatas();
static void setMonths(), selectCurrentDay();
static void listAppoints(), checkAndSaveAppoints();
static int countAppoints();
static void cvtStringToPixels(), cvtStringToPixmaps(), syntax();
static Pixmap readBitmap();

/*
 * Global graphics data
 */
Display *display;
Screen *screen;
Window root;
GC dateGC1,dateGC2,dateGC3,dateGC12,emptyGC,*shadeGC;

/*
 * Global widget data
 */
XtAppContext app_con;
Widget toplevel;
Widget bothForm,allForm;
AppResources appResources;
MonthFormData *monthFormData1,*monthFormData2[2];
MonthFormData *monthFormData3[3],*monthFormData12[12];

/*
 * Other global data
 */
char *program;
int currentDay,currentMon,currentYear;		/* today */
int mainDay,mainMon,mainYear;			/* display */
int appointsChanged = False;			/* need to save? */
XtIntervalId timeoutId;				/* auto update */

/*
 * Action binding table
 */
static XtActionsRec cmdActionsTable[] = {
	{ "xkal-quit", quit },
	{ "xkal-exit", quit_nosave },
	{ "xkal-save", save },
	{ "xkal-next", next },
	{ "xkal-prev", prev },
	{ "xkal-today", today },
	{ "xkal-numMonths", setNumMonths },
	{ "xkal-system", escapeToSystem },
	{ "xkal-focus-this-item", focusThisItem },
	{ "xkal-focus-no-item", focusNoItem },
	{ "xkal-focus-next-item", focusNextItem },
	{ "xkal-focus-prev-item", focusPrevItem },
	{ "xkal-edit-appoint", editAppoint },
	{ "xkal-focus-next-edit-item", focusNextEditItem },
	{ "xkal-focus-prev-edit-item", focusPrevEditItem },
	{ "xkal-edit-defaults", editDefaults },
	{ "xkal-focus-next-defaults-item", focusNextDefaultsItem },
	{ "xkal-focus-prev-defaults-item", focusPrevDefaultsItem },
};

/*
 *	Widget and non-widget resources if the application defaults
 *	file can't be found.
 *	[ Generated automatically from Xkal.ad. ]
 */
static String fallbackResources[] = {
#include "Xkal.ad.h"
	NULL
};

#define ACTION_PROC(NAME)	void NAME(w,event,params,num_params) \
					Widget w; \
					XEvent *event; \
					String *params; \
					Cardinal *num_params;

/*	-	-	-	-	-	-	-	-	*/

int
main(argc,argv)
int argc;
char **argv;
{
    int i;

    program = *argv;
    for (i=1; i < argc; i++)		/* quick check for non-interactive */
	if (strncmp(argv[i],"-l",2) == 0 || strncmp(argv[i],"-s",2) == 0 ||
						strncmp(argv[i],"-v",2) == 0)
	    break;
    if (i != argc)
	getResources(&argc,argv);	/* found, so don't need display */
    else
	initGraphics(&argc,argv);	/* not found, so interactive mode */
    if (argc > 1) {
	syntax(argc,argv);
	XtDestroyApplicationContext(app_con);
	exit(-1);
    }
    initDb();
    initDateStrings();
    getCurrentDate(&currentDay,&currentMon,&currentYear);
    if (appResources.systemAppoints && *appResources.systemAppoints)
	readDb(appResources.systemAppoints,True);
    if (appResources.personalAppoints && *appResources.personalAppoints)
	readDb(appResources.personalAppoints,False);
    if (appResources.date != NULL && *(appResources.date) != '\0') {
	parseDate(appResources.date,&currentDay,&currentMon,&currentYear);
	if (currentYear < 1901) {
	    fprintf(stderr,"%s: invalid year (too early): %d\n",program,
								currentYear);
	    XtDestroyApplicationContext(app_con);
	    exit(-1);
	}
    }
    if (appResources.version) {
	printf("xkal %.2f - George Ferguson (ferguson@cs.rochester.edu)\n",
								XKAL_VERSION);
	exit(0);
    }
    if (appResources.listOnly) {
	listAppoints(currentDay,currentMon,currentYear);
    }
    if (appResources.listOnly || appResources.silent) {
	exit(countAppoints(currentDay,currentMon,currentYear,
						 appResources.exitUsesLevels));
    }
    mainDay = currentDay;
    mainMon = currentMon;
    mainYear = currentYear;
    initGCs();
    initWidgets();
    XtRealizeWidget(toplevel);
    if (currentDayFormData != NULL)
	selectCurrentDay();
    if (appResources.checkpointInterval > 0)
	timeoutId = XtAppAddTimeOut(app_con,
			(unsigned long)(appResources.checkpointInterval*60000),
			timeoutProc,NULL);
    XtAppMainLoop(app_con);
}

/*	-	-	-	-	-	-	-	-	*/
/* Initialization routines */

/*
 * initGraphics() : Initialze X connection and toplevel widget.
 */
static void
initGraphics(argcp,argv)
int *argcp;
char **argv;
{
    char *rev,*strchr();

    toplevel = XtAppInitialize(&app_con, "Xkal",
			       xkalOptions, XtNumber(xkalOptions),
			       argcp,argv,fallbackResources,NULL,ZERO);
    XtGetApplicationResources(toplevel,(XtPointer)&appResources,
                              xkalResources,XtNumber(xkalResources),NULL,ZERO);
    rev = strchr(fallbackResources[0],'$');
    if (rev && strcmp(appResources.revision,rev) != 0) {
	fprintf(stderr,"%s: app-defaults release not %s\n",program,rev);
	fprintf(stderr,"%s: you may have an outdated app-defaults file\n",
								program);
    }
    XtAppAddActions(app_con,cmdActionsTable,XtNumber(cmdActionsTable));
    display = XtDisplay(toplevel);
    screen = XtScreen(toplevel);
    root = RootWindowOfScreen(screen);
}

/*
 * initGCs() : Initialize GC's for drawing operations.
 */
static void
initGCs()
{
    XGCValues values;
    XtGCMask mask;
    Pixel *pixels;
    Pixmap *pixmaps;
    int i;

    /*
     * initialize date GC's
     */
    values.foreground = (unsigned long)1;
    values.background = (unsigned long)0;
    values.function = GXset;
    mask = GCForeground | GCBackground | GCFunction | GCFont;
    values.font = appResources.dateFont1;
    dateGC1 = XCreateGC(display,root,mask,&values);
    values.font = appResources.dateFont2;
    dateGC2 = XCreateGC(display,root,mask,&values);
    values.font = appResources.dateFont3;
    dateGC3 = XCreateGC(display,root,mask,&values);
    values.font = appResources.dateFont12;
    dateGC12 = XCreateGC(display,root,mask,&values);
    /*
     * initialize empty GC
     */
    values.function = GXcopy;
    values.fill_style = FillTiled;
    values.tile = readBitmap(appResources.noDayShade);
    mask = GCForeground | GCBackground | GCFunction | GCFillStyle | GCTile;
    emptyGC = XCreateGC(display,root,mask,&values);
    /*
     * initialize shading GC's
     */
    values.function = GXcopy;
    mask = GCFunction;
    shadeGC = (GC *)XtCalloc(appResources.maxLevel+1,sizeof(GC));
    shadeGC[0] = XCreateGC(display,root,mask,&values);
    if (DefaultDepthOfScreen(screen) > 1) {	/* COLOR */
	pixels = (Pixel *)XtCalloc(appResources.maxLevel,sizeof(Pixel));
	cvtStringToPixels(appResources.colors,appResources.maxLevel,pixels);
	for (i=1; i <= appResources.maxLevel; i++) {
	    values.foreground = pixels[i-1];
	    values.fill_style = FillSolid;
	    values.function = GXcopy;
	    mask = GCForeground | GCBackground | GCFunction;
	    shadeGC[i] = XCreateGC(display,root,mask,&values);
	}
	XtFree(pixels);
    } else {					/* B&W */
	pixmaps = (Pixmap *)XtCalloc(appResources.maxLevel,sizeof(Pixmap));
	cvtStringToPixmaps(appResources.shades,appResources.maxLevel,pixmaps);
	for (i=1; i <= appResources.maxLevel; i++) {
	    values.background = (unsigned long)0;
	    values.tile = pixmaps[i-1];
	    values.fill_style = FillTiled;
	    values.function = GXcopy;
	    mask = GCForeground | GCBackground | GCTile | GCFillStyle |
								GCFunction;
	    shadeGC[i] = XCreateGC(display,root,mask,&values);
	}
	XtFree(pixmaps);
    }
}

/*
 * initWidgets() : Create necessary widgets and menus (ie. those that
 *	aren't created on the fly).
 */
static void
initWidgets()
{
    bothForm = XtCreateManagedWidget("bothForm",formWidgetClass,toplevel,
								NULL,ZERO);
    XtCreateManagedWidget("fileButton",menuButtonWidgetClass,bothForm,
								NULL,ZERO);
    XtCreatePopupShell("fileMenu",ezMenuWidgetClass,bothForm,NULL,ZERO);
    XtCreateManagedWidget("viewButton",menuButtonWidgetClass,bothForm,
								NULL,ZERO);
    XtCreatePopupShell("viewMenu",ezMenuWidgetClass,bothForm,NULL,ZERO);
    XtCreateManagedWidget("otherButton",menuButtonWidgetClass,bothForm,
								NULL,ZERO);
    XtCreatePopupShell("otherMenu",ezMenuWidgetClass,bothForm,NULL,ZERO);
    allForm = XtCreateManagedWidget("allMonthsForm",formWidgetClass,bothForm,
								NULL,ZERO);
    createMonthFormDatas(appResources.numMonths);
    if (appResources.bothShown)
	currentDayFormData = createDayFormData(bothForm);
    setMonths();
}

/*
 * createMonthFormDatas(n) : Create n monthForms as children of the
 *	allForm widget.
 */
static void
createMonthFormDatas(n)
int n;
{
    char *name = "monthFormXX_XX";
    int i;

    switch (n) {
	case 1: monthFormData1 = createMonthFormData(allForm,"monthForm1",1);
		XtManageChild(monthFormData1->form);
		break;
	case 2: for (i=0; i < 2; i++) {
		    sprintf(name,"monthForm2_%d",i+1);
		    monthFormData2[i] = createMonthFormData(allForm,name,2);
		    XtManageChild(monthFormData2[i]->form);
		}
		break;
	case 3: for (i=0; i < 3; i++) {
		    sprintf(name,"monthForm3_%d",i+1);
		    monthFormData3[i] = createMonthFormData(allForm,name,3);
		    XtManageChild(monthFormData3[i]->form);
		}
		break;
	case 12: for (i=0; i < 12; i++) {
		    sprintf(name,"monthForm12_%d",i+1);
		    monthFormData12[i] = createMonthFormData(allForm,name,12);
		    XtManageChild(monthFormData12[i]->form);
		}
		break;
    }
}

/*
 * setMonths() : Redraw the months.
 */
static void
setMonths()
{
    int i,prevMon,prevYear,nextMon,nextYear;

    XawFormDoLayout(allForm,False);
    switch (appResources.numMonths) {
	case 1:  setMonthFormData(monthFormData1,1,mainMon,mainYear);
		 break;
	case 2:	 if (mainMon == 12) {
		     nextMon = 1;
		     nextYear = mainYear+1;
		 } else {
		     nextMon = mainMon+1;
		     nextYear = mainYear;
		 }
		 setMonthFormData(monthFormData2[0],2,mainMon,mainYear);
		 setMonthFormData(monthFormData2[1],2,nextMon,nextYear);
		 break;
	case 3:  prevMon = mainMon - 1;
		 nextMon = mainMon + 1;
		 prevYear = nextYear = mainYear;
		 if (mainMon == 1) {
		     prevMon = 12;
		     prevYear = mainYear - 1;
		 } else if (mainMon == 12) {
		     nextMon = 1;
		     nextYear = mainYear + 1;
		 }
		 setMonthFormData(monthFormData3[0],3,prevMon,prevYear);
		 setMonthFormData(monthFormData3[1],3,mainMon,mainYear);
		 setMonthFormData(monthFormData3[2],3,nextMon,nextYear);
		 break;
	case 12: for (i=0; i < 12; i++)
		     setMonthFormData(monthFormData12[i],12,i+1,mainYear);
		 break;
    }
    XawFormDoLayout(allForm,True);
}

/*
 * selectCurrentDay() : Higlight current day after figuring out which
 *	form its in.
 */
static void
selectCurrentDay()
{
    switch (appResources.numMonths) {
	case 1: selectDay(monthFormData1,mainDay,mainMon,mainYear);
		break;
	case 2: selectDay(monthFormData2[0],mainDay,mainMon,mainYear);
		break;
	case 3: selectDay(monthFormData3[1],mainDay,mainMon,mainYear);
		break;
	case 12: selectDay(monthFormData12[mainMon-1],mainDay,mainMon,mainYear);
		 break;
    }
}

/*	-	-	-	-	-	-	-	-	*/

/*
 * quit() : Save appoints if changed, then quit.
 */
/*ARGSUSED*/
ACTION_PROC(quit)
{
    checkAndSaveAppoints(False);
    XtDestroyApplicationContext(app_con);
    exit(0);
}

/*
 * quit_nosave() : Quit without saving.
 */
/*ARGSUSED*/
ACTION_PROC(quit_nosave)
{
    XtDestroyApplicationContext(app_con);
    exit(0);
}

/*
 * next() : Display next month.
 */
/*ARGSUSED*/
ACTION_PROC(next)
{
    if (appResources.numMonths < 12) {
	if (mainMon == 12) {
	    mainMon = 1;
	    mainYear += 1;
	} else {
	    mainMon += 1;
	}
    } else {
	mainYear += 1;
    }
    setMonths();
}

/*
 * prev() : Display previous month.
 */
/*ARGSUSED*/
ACTION_PROC(prev)
{
    if (appResources.numMonths < 12) {
	if (mainMon == 1) {
	    mainMon = 12;
	    mainYear -= 1;
	} else {
	    mainMon -= 1;
	}
    } else {
	mainYear -= 1;
    }
    setMonths();
}

/*
 * today() : Highlight today, possibly resetting months. Also, if date
 *	was not given as resource, then get it again in case we've been
 *	runnign for a while.
 */
/*ARGSUSED*/
ACTION_PROC(today)
{
    /* Reset the current date in case we've been running for a long time. */
    if (appResources.date == NULL || *(appResources.date) == '\0')
	getCurrentDate(&currentDay,&currentMon,&currentYear);
    mainDay = currentDay;
    if (mainMon != currentMon || mainYear != currentYear) {
	mainMon = currentMon;
	mainYear = currentYear;
	setMonths();
    }
    selectCurrentDay();
}

/*
 * setNumMonths(n) : Change the number of months being displayed, possibly
 *	creating new monthForms. n should be 1, 2, 3, or 12.
 */
/*ARGSUSED*/
ACTION_PROC(setNumMonths)
{
    int n,i;

    if (*num_params != ONE) {
	fprintf(stderr,"%s: bad parms to xkal-numMonths()\n",program);
	return;
    }
    n = atoi(params[0]);
    if (n != 1 && n != 2 && n != 3 && n != 12) {
	fprintf(stderr,"%s: bad num for xkal-numMonths(): %d\n",program,n);
	return;
    }
    if (n == appResources.numMonths)
	return;
    XawFormDoLayout(bothForm,False);
    switch (appResources.numMonths) {
	case 1: XtUnmanageChild(monthFormData1->form);
		break;
	case 2: for (i=0; i < 2; i++)
		    XtUnmanageChild(monthFormData2[i]->form);
		break;
	case 3: for (i=0; i < 3; i++)
		    XtUnmanageChild(monthFormData3[i]->form);
		break;
	case 12: for (i=0; i < 12; i++)
		    XtUnmanageChild(monthFormData12[i]->form);
		break;
    }
    appResources.numMonths = n;
    switch (appResources.numMonths) {
	case 1: if (monthFormData1 == NULL)
		    createMonthFormDatas(1);
		setMonths();
		XtManageChild(monthFormData1->form);
		break;
	case 2: if (monthFormData2[0] == NULL)
		    createMonthFormDatas(2);
		setMonths();
		for (i=0; i < 2; i++)
		    XtManageChild(monthFormData2[i]->form);
		break;
	case 3: if (monthFormData3[0] == NULL)
		    createMonthFormDatas(3);
		setMonths();
		for (i=0; i < 3; i++)
		    XtManageChild(monthFormData3[i]->form);
		break;
	case 12: if (monthFormData12[0] == NULL)
		     createMonthFormDatas(12);
		 setMonths();
		 for (i=0; i < 12; i++)
		     XtManageChild(monthFormData12[i]->form);
		 break;
    }
    XawFormDoLayout(bothForm,True);
}

/*
 * save() : Save appoints regardless of whether they've changed or not.
 */
/*ARGSUSED*/
ACTION_PROC(save)
{
    checkAndSaveAppoints(True);
}

/*
 * escapeToSystem(str) : execute "str" by the shell.
 */
/*ARGSUSED*/
ACTION_PROC(escapeToSystem)
{
    if (*num_params != 1) {
#ifdef USE_ALERT
	alert("Error: too many arguments to xkal-system().");
#else
	fprintf(stderr,"\007%s: too many arguments to xkal-system()\n",program);
#endif
	return;
    }
    system(params[0]);
}

/*
 * focusThisItem() : Force input focus to this widget.
 */
/*ARGSUSED*/
ACTION_PROC(focusThisItem)
{
    XWindowAttributes attrs;

    XGetWindowAttributes(display,XtWindow(w),&attrs);
    if (attrs.map_state != IsViewable)
	return;
    XSetInputFocus(display,XtWindow(w),RevertToParent,CurrentTime);
}

/*
 * focusNoItem() : Unset the input focus.
 */
/*ARGSUSED*/
ACTION_PROC(focusNoItem)
{
    XSetInputFocus(display,PointerRoot,RevertToParent,CurrentTime);
}

/*
 * focusNextItem() : Switch input focus to next appointment slot, wrapping
 *	at bottom.
 */
/*ARGSUSED*/
ACTION_PROC(focusNextItem)
{
    int i;

    for (i=0; i < numDisplayedAppoints; i++)
	if (w == currentDayFormData->items[i]->text)
	    break;
    if (i == numDisplayedAppoints)
	return;
    else if (i == numDisplayedAppoints-1)
	i = 0;
    else
	i += 1;
    XSetInputFocus(display,XtWindow(currentDayFormData->items[i]->text),
						RevertToParent,CurrentTime);
}

/*
 * focusPrevItem() : Switch input focus to previous appointment slot, wrapping
 *	at top.
 */
/*ARGSUSED*/
ACTION_PROC(focusPrevItem)
{
    int i;

    for (i=0; i < numDisplayedAppoints; i++)
	if (w == currentDayFormData->items[i]->text)
	    break;
    if (i == numDisplayedAppoints)
	return;
    else if (i == 0)
	i = numDisplayedAppoints-1;
    else
	i -= 1;
    XSetInputFocus(display,XtWindow(currentDayFormData->items[i]->text),
						RevertToParent,CurrentTime);
}

/*
 * timeoutProc() : Timer callback - saves appoints if they've changed and
 *	resets timer.
 */
/*ARGSUSED*/
void
timeoutProc(data,id)
XtPointer data;
XtIntervalId *id;
{
    checkAndSaveAppoints(False);
    if (appResources.checkpointInterval > 0)
	timeoutId = XtAppAddTimeOut(app_con,
			(unsigned long)(appResources.checkpointInterval*60000),
			timeoutProc,NULL);
}

/*	-	-	-	-	-	-	-	-	*/
/* Miscellaneous routines */

/*
 * listAppoints() : Print all appoints for day to stdout.
 */
static void
listAppoints(day,mon,year)
int day,mon,year;
{
    Msgrec **apps;
    int dow,n,i;

    dow = computeDOW(day,mon,year);
    printf("%s %d %s %d\n",shortDowStr[dow-1],day,shortMonthStr[mon-1],year);
    n = lookupEntry(day,mon,year,-1,-1,-1,NULL,False);
    if (n == 0)
	return;
    apps = (Msgrec **)XtCalloc(n,sizeof(Msgrec *));
    (void)lookupEntry(day,mon,year,-1,-1,n,apps,False);
    for (i=0; i < n; i++)
	if (apps[i]->hour == -1)
	    printf("\t%s %s\n",appResources.notesLabel,apps[i]->text);
	else if (apps[i]->mins == -1)
	    if (apps[i]->hour >= 12)
		printf("\t%2dpm %s\n",apps[i]->hour-12,apps[i]->text);
	    else
		printf("\t%2dam %s\n",apps[i]->hour,apps[i]->text);
	else
	    printf("\t%2d:%02d %s\n",apps[i]->hour,apps[i]->mins,apps[i]->text);
    XtFree(apps);
}

/*
 * countAppoints() : Return number of appoints (flag = False) or criticality
 *	total (flag = True) for day.
 */
static int
countAppoints(day,mon,year,flag)
int day,mon,year;
Boolean flag;
{
    
    return(lookupEntry(day,mon,year,-1,-1,-1,NULL,flag));
}

/*
 * checkAndSaveAppoints() : Transfer data from dayForm to DB, then save
 *	appoints if they've changed.
 */
static void
checkAndSaveAppoints(force)
Boolean force;
{
    if (currentDayFormData != NULL)
	checkpointAppoints(currentDayFormData);
    if (force || appointsChanged)
	writeDb(appResources.personalAppoints,False);
    appointsChanged = False;
}

/*	-	-	-	-	-	-	-	-	*/
/* Miscellaneous conversion routines */

/*
 * cvtStringToPixels() : Converts a string that is a whitespace-separated
 *	list of pixel names, to an array of Pixels.
 */
static void
cvtStringToPixels(str,max,pixels)
char *str;
int max;
Pixel *pixels;
{
    XColor screen_in_out,exact;
    char buf[64];
    int i,j;

    strcpy(buf,XtDefaultBackground);
    for (i=0; i < max; i++) {
	if (*str) {
	    while (isspace(*str))
		str += 1;
	    if (!*str)
		break;
	    j = 0;
	    while (*str && j < 63 && !isspace(*str))
		buf[j++] = *str++;
	    buf[j] = '\0';
	    if (*str && !isspace(*str)) {
		fprintf(stderr,"%s: color name too long: %s\n",program,buf);
		while (*str && !isspace(*str))
		    str += 1;
	    }
	}
	XAllocNamedColor(display,DefaultColormapOfScreen(screen),buf,
							&screen_in_out,&exact);
	pixels[i] = screen_in_out.pixel;
    }
}

/*
 * cvtStringToPixmaps() : Converts a string that is a whitespace-separated
 *	list of bitmap names, to an array of Pixmaps.
 */
static void
cvtStringToPixmaps(str,max,pixmaps)
char *str;
int max;
Pixmap *pixmaps;
{
    char buf[64];
    int i,j;

    strcpy(buf,"gray");
    for (i=0; i < max; i++) {
	if (*str) {
	    while (isspace(*str))
		str += 1;
	    if (!*str)
		break;
	    j = 0;
	    while (*str && j < 63 && !isspace(*str))
		buf[j++] = *str++;
	    buf[j] = '\0';
	    if (*str && !isspace(*str)) {
		fprintf(stderr,"%s: shade name too long: %s\n",program,buf);
		while (*str && !isspace(*str))
		    str += 1;
	    }
	}
	pixmaps[i] = readBitmap(buf);
    }
}

/*
 * readBitmap() : Reads the given file using XmuLocateBitmapFile protocol,
 *	then converts to Pixmap and returns it. If not found, prints error
 *	message and returns a default Pixmap.
 */
static Pixmap
readBitmap(name)
char *name;
{
    Pixmap b;
    int w,h,xhot,yhot;

    b = XmuLocateBitmapFile(screen,name,NULL,ZERO,&w,&h,&xhot,&yhot);
    if (b == NULL) {
	fprintf(stderr,"%s: couldn't find bitmap \"%s\"\n",program,name);
	return(XCreatePixmap(display,screen,2,2,DefaultDepthOfScreen(screen)));
    } else
	return(XmuCreatePixmapFromBitmap(display,root,b,w,h,
					DefaultDepthOfScreen(screen),1,0));
}

/*	-	-	-	-	-	-	-	-	*/
/*
 * syntax() : Print whatever caused the error and the usage message.
 */
static void
syntax(argc,argv)
int argc;
char **argv;
{
    argv += 1;
    if (argc > 2 || (strcmp(*argv,"-help") != 0 && strcmp(*argv,"-?") != 0)) {
	fprintf(stderr,"%s: bad argument(s): ",program);
	while (--argc)
	    fprintf(stderr,"%s ",*argv++);
	fprintf(stderr,"\n");
    }
    fprintf(stderr,"usage: %s",program);
    fprintf(stderr,"\t-appoints file\tuse file for personal appointments\n");
    fprintf(stderr,"\t\t-date date\tuse given date for `today'\n");
    fprintf(stderr,"\t\t-numMonths 1|2|3|12\tstart displaying 1, 2, 3 or 12 months\n");
    fprintf(stderr,"\t\t-bothShown\tput day display next to month(s)\n");
    fprintf(stderr,"\t\t-nobothShown\tpopup days as needed\n");
    fprintf(stderr,"\t\t-version\tprint version info to stdout\n");
    fprintf(stderr,"\t\t-listOnly\tprint appoints to stdout\n");
    fprintf(stderr,"\t\t-silent\t\texit with status == number of appoints\n");
    fprintf(stderr,"\t\t-exitUsesLevels\texit status is criticality total\n");
    fprintf(stderr,"\t\t-noexitUsesLevels\texit status is number of appoints\n");
    fprintf(stderr,"\t\t-opaqueDates\tprint day numbers opaquely\n");
    fprintf(stderr,"\t\t-noopaqueDates\tprint day numbers transparently\n");
    fprintf(stderr,"\t\t-dowLabels\tprint day of week at top of columns\n");
    fprintf(stderr,"\t\t-nodowLabels\tdon't print dow at top of columns\n");
    fprintf(stderr,"\t\t-titlebar\tuse titlebars to display dates\n");
    fprintf(stderr,"\t\t-notitlebar\tdon't use titlebars to display dates\n");
    fprintf(stderr,"\t\t-checkpointInterval mins\tmins between autosaves\n");
    fprintf(stderr,"\t\t-xrm 'resource: value'\tpass arbitrary resources\n");
}
